### ISSUE TITLE
add support for SoundCloud

### ISSUE THREAD

**Issue by HumbleBeeBumbleBee** [tgwizard/sls#144]
Support for SoundCloud scrobble.

**Comment by inverse** [comment]
I've managed to figure out that the SoundCloud application actually uses

- com.android.music.playstatechanged
- com.android.music.metachanged

**Comment by vedit** [comment]
Is there any ongoing development here? I'd like to pitch in if nobody built something so far.

**Comment by inverse** [comment]
I started working a little bit on a branch (see https://github.com/inverse/sls/tree/soundcloud). 

- Need to figure out the correct parsing of the track/artist since often the track contains the artist as well. This is also apparent with reposts where the artist bundle key is the reposter, and not the actual artist. I can see this causing quite a lot of invalid submissions.

- Also there may be another intent that I missing (`com.android.music.playbackcomplete`) which needs to be investigated

- Finally working out if there is any further information in the playstate change intent.


I'm really unsure as to if anything will ever get merged here @tgwizard has been pretty absent this past year.

**Comment by inverse** [comment]
@vedit feel free to contribute or take the work what I've done to bring SC integration into the player :)

**Comment by vedit** [comment]
@inverse thanks, I'll take a look at it today

**Comment by jeffshieldsdev** [comment]
Any update on Soundcloud support? Thank you.

**Comment by vedit** [comment]
Sorry for lack of an update. I took a stab at it and turns out correct names are the main problem with soundcloud. Couldn't reliably find and use track names for lastfm submissions, especially from record label pages.

**Comment by bhb192** [comment]
There is no way you will ever be able to automatically correct every single song on Soundcloud. It's better to just scrobble is as-is because then everybody will be scrobbling the same thing, even if it's technically "wrong." If 3000 people scrobble a song with incorrect info, last.fm will be much more likely to file a correction than 20 scrobbles on a "fixed" track title from this specific app.

**Comment by inverse** [comment]
Perhaps looking at how the [web-scrobbler][0] handles Soundcloud for some ideas of how the track scrobbling could be done.

[0]: https://github.com/david-sabata/web-scrobbler/blob/master/connectors/soundcloud.js

**Comment by HumbleBeeBumbleBee** [comment]
@inverse I found another resource for you on parsing.
https://addons.mozilla.org/en-US/firefox/files/browse/304994/file/chrome/content/foxyscrobbler/soundcloud.js
Java isn't my strong suit, but I understand these web-scrobblers and can understand most common programming languages. I hope this helps you with your sls/soundcloud fork.

**Comment by inverse** [comment]
@HumbleBeeBumbleBee  Thanks for the heads up. I think It'll be good to look at how these other web-scrobblers handle such cases because It's not going to be an absolute solution but def an improvement over not scrobbling. 

Also we could add some validation against last.fm to only scrobble known tracks. But I am not sure what people's take on that would be.

**Comment by bhb192** [comment]
I've already said my take on the issue, but just to reiterate...I think we
should only worry about scrobbling the track info as-is. So Artist Name -
Name of Track. Then let the last.FM community decide what things need to be
corrected to.

Everybody that uses soundcloud knows how weird some of the track names can
be, especially when you have one artist posting something where the primary
artist is actually in the track title. So you end up with Artist A - Artist
B - Track Name. Big deal and so what? If that's how they want it to show up
then just let it be and stop worrying about trying to make a panacea for
every situation.

**Comment by HumbleBeeBumbleBee** [comment]
I have a working prototype of SoundCloud running on my phone now.
No parsing to the data.
I second the motion from @bhb192 about letting last.fm and libre.fm to decide what to do with the songs.
It scrobbles just fine, except I have found that it cuts of Track names that exceed a certain size.
`Mr. Little Jeans - The Suburbs (Crywolf`
Soundcloud hands over a shortened track name according to my logs.
```
	Line 705447: E/SoundCloud(27592): 411 <- track length in seconds for @ 50% scrobble
	Line 705449: E/SoundCloud(27592): Dumping Intent start
	Line 705451: E/SoundCloud(27592): [duration=411689]
	Line 705453: E/SoundCloud(27592): [artist=Crywolf]
	Line 705455: E/SoundCloud(27592): [id=27701972]
	Line 705457: E/SoundCloud(27592): [track=Mr. Little Jeans - The Suburbs (Crywolf ]
```
I don't think this is an SLS probably but a SoundCloud problem.

**Comment by bhb192** [comment]
Thanks for all of your work on that, I'm happy to see it finally
scrobbling! What is the data source of the track name?

**Comment by HumbleBeeBumbleBee** [comment]
I second the motion because SoundCloud doesn't send the full track information anyways.
`Mr. Little Jeans - The Suburbs (Crywolf` <- only about 40 characters
It would be different if full information was in the track title.
If that were so we could split the title into `track name - artist` and maybe have a Spotify style comma separated artist and track information.

**Comment by HumbleBeeBumbleBee** [comment]
@bhb192 The data source of the track is com.android.music.metachanged
I go that from @inverse

**Comment by bhb192** [comment]
When I use SoundCloud with Bluetooth it sends the artist and the track name
separately, so maybe there is a way to get more characters if that can be
accessed.

**Comment by HumbleBeeBumbleBee** [comment]
@bhb192  I fired out and email to help@soundcloud.com
I'll let you know if there are any custom broadcasts I can grab full track information from. :)

**Comment by HumbleBeeBumbleBee** [comment]
If there isn't going to be any parsing involved I guess we can at least try to get all the information to last.fm or libre.fm so they can fix track titles.

**Comment by HumbleBeeBumbleBee** [comment]
https://github.com/tgwizard/sls/blob/SoundCloud/src/com/adam/aslfms/receiver/SoundCloud.java
Ok, it's all done. The output is messy, but we have moderate SoundCloud support now.

**Comment by inverse** [comment]
The cut of in intent track info seems a little strange. Let us know what SoundCloud say :)

Is it worth porting some of the logic from web-scrobbler to clean up the track information? I guess that could be useful to have built into the core and enabled on a per-player basis. Perhaps for now just bundling it with SoundCloud would be sufficient for now. Opinions?

Why do you say scrobbler after 4 minutes?

**Comment by HumbleBeeBumbleBee** [comment]
A track should only be scrobbled when the following conditions have been met:

 * The track must be longer than 30 seconds.
 * And the track has been played for at least half its duration, or for 4 minutes (whichever occurs earlier.)

http://www.last.fm/api/scrobbling

**Comment by inverse** [comment]
Makes sens :+1:  I'd say it'll be worth raising a separate issue to discuss how this should be handled.

**Comment by bhb192** [comment]
If you ever scrobbled hour-long DJ mixes you'll notice that they scrobble
way before the half way point, hence the four minutes.

The reason I am so adamant about not doing a track cleanup is because it
will never be a catch-all and you actually end up scrobbling unique tracks
which are still incorrect, instead of ones that we know are wrong but still
have thousands of scrobbles already. Just look at the artist page for
octobersveryown.

Thanks again for all of your efforts, they are very much appreciated!

**Comment by HumbleBeeBumbleBee** [comment]
@bhb192, so you can report the 4 minute rule working?

**Comment by bhb192** [comment]
Not from this usage specifically, but the computer-based scrobbler has done
that (or used to do that) since at least 2007.

**Comment by HumbleBeeBumbleBee** [comment]
I didn't realize there was an alternative out there...
[My Cloud Player](http://mycloudplayers.com/) ([Google Play](https://play.google.com/store/apps/details?id=com.mycloudplayers.mycloudplayer))

**Comment by bhb192** [comment]
Yeah and it works but the main reason I stopped using it was because
SoundCloud banned caching by third party apps and I need that to be able to
use it when I don't have service.

**Comment by HumbleBeeBumbleBee** [comment]
@inverse did you want me to branch soundcloud for you and let you make the pull request or just merge it?

**Comment by inverse** [comment]
@HumbleBeeBumbleBee I'm happy for you to do it :+1:

**Comment by HumbleBeeBumbleBee** [comment]
Thanks, and sorry, I suck at Git.

**Comment by HumbleBeeBumbleBee** [comment]
Ok all good :)

**Comment by inverse** [comment]
Just be careful with git! Can be accidentally destructive if you run the wrong command ;)

**Comment by HumbleBeeBumbleBee** [comment]
Thanks for reminding me. :) :+1: 
I am using Windows and it's kind of terrible because the default text editor is notepad.exe
I don't usually use the command-line for that reason.
I have been using manual editing with the website and copy paste into notepad for changes in gitshell.

I will have to setup a Debian OS later and get comfortable with git commands.

**Comment by inverse** [comment]
You don't need debian or linux to get comfortable with git commands. Also I'd recommend using notepad2 as a drop in replacement for the default windows notepad.

Also It might be worth checking out

- SourceTree - decent standalone git client
- tortoisegit - nicely integrated with explorer
- conemu
- chocolatey - almost like a package manager for windows and a nice way to install apps :+1: 


... anyway i'm going off topic ;)

**Comment by HumbleBeeBumbleBee** [comment]
Thanks.

**Comment by writepython** [comment]
Regarding data cleanup.  You could try a few things first, and then fall back to just scrobbling the raw data.  If you're getting the soundcloud name, you could first check last.fm's api's artist autocorrection feature and see if that returns an artist name.  If that fails, you can split on "-" and see if last.fm's api's artist autocorrection feature returns an artist name for either side of that split.  And finally, give up and scrobble the raw data.

**Comment by bhb192** [comment]
If you ask last.fm if the artist/track already exists, most of the time it
will because users from 8tracks have already scrobbled the raw data.

**Comment by writepython** [comment]
@bhb192 - If you're saying that 8tracks and others indiscriminately scrobble soundcloud username as artist name, that does complicate things.  I see evidence of that here: http://www.last.fm/music/TriAngleRecords   In that case, it might be better to split the track name first on "-" and see if last.fm's api's artist autocorrection feature returns an artist name for either side of that split.  And look at souncloud username second.

**Comment by bhb192** [comment]
Most of the time making the SoundCloud username second priority will work,
but then there are times when the artist has a dash in the name. Also a
dash isn't always the delineator. I mentioned it in a previous post, but
look at http://last.fm/music/octobersveryown. And that is the official
account!

**Comment by writepython** [comment]
Right.  The splitting could be a bit more robust.  Search the string for: whitespace + some assortment of delimeters + whitespace.

**Comment by HumbleBeeBumbleBee** [comment]
https://stackoverflow.com/questions/30083644/what-causes-soundcloud-track-strings-character-limits

**Comment by HumbleBeeBumbleBee** [comment]
I have a new issue with SoundCloud becoming the default receiver instead of the built in music player. I think this could be the problem affecting 8tracks and VLC and my stock player not scrobbling properly.
I might have to drop SoundCloud support until this is fixed, but I want to make sure it's the cause of the problem.

**Comment by HumbleBeeBumbleBee** [comment]
It's 100% caused by [SoundCloudReceiver.java](https://github.com/tgwizard/sls/blob/master/app/src/main/java/com/adam/aslfms/receiver/SoundCloudReceiver.java)

When "SoundCloud" is enabled in "Enabled Apps" [VLC](https://www.videolan.org/vlc/download-android.html), Stock Samsung Music Player and [8tracks](https://8tracks.com/) DO NOT scrobble to last.fm, they only scrobble to libre.fm.

When "SoundCloud" is disabled in "Enabled Apps" everything works normally.

I believe the conflict is here with
https://github.com/tgwizard/sls/blob/master/app/src/main/AndroidManifest.xml#L263-L271
overriding/interfering
https://github.com/tgwizard/sls/blob/master/app/src/main/AndroidManifest.xml#L67-L76

I am going to try switching the places of the above code, but I still think they will conflict.

I think this is going to require asking SoundCloud Android developers to make their application Scrobbler compatible.

**Comment by bhb192** [comment]
Not a solution to the problem, but 8tracks will scrobble server-side. I
wish SoundCloud would just give an actual statement regarding scrobbling
other than "we don't support it, but sharing to Facebook is similar."

**Comment by inverse** [comment]
I don't think that having two broadcast receivers listen to the same intent would cause a problem. Surely you can have multiple receivers in the same application listening to the same intents. 

Perhaps we need to add a check to the Soundcloud one to check that the intent is from the right package. Are we able to do that? It's been a while since I've worked with Android and my phone's USB is dodgy

**Comment by HumbleBeeBumbleBee** [comment]
I left a branch [OldSC](https://github.com/tgwizard/sls/tree/OldSoundCloud) with the current SoundCloud support as it is for people who want to build there own a manually enable/disable SoundCloud as they need it, but it seems like there are too many issues. 
* 40 character limit
* default android intent interference
* patch that limits our device support to KitKat (API 19+) [Version Graph](https://upload.wikimedia.org/wikipedia/commons/e/ee/Android_historical_version_distribution_-_vector.svg)

I can't cut of (potentially) half our user base for bad SoundCloud support.
If somebody wants to try and patch this on [OldSC](https://github.com/tgwizard/sls/tree/OldSoundCloud) please do, but I want to keep the support as it's shown on Google Play
* Requires Android 2.1 and up

We should just have a bunch of people demanding Scrobble support [help@soundcloud.com](mailto:help@soundcloud.com)
For now I will put SoundCloud in the [PBR category](https://github.com/tgwizard/sls/blob/master/IssueSummary.md#apps).

**Comment by bhb192** [comment]
I've given up on SoundCloud and decided to just get the SoundLoader app
from the Amazon store. This way, I don't have to rely on caching and can
play/scrobble everything from the stock Music app. Problem solved!

**Comment by HumbleBeeBumbleBee** [comment]
@bhb192 Great!!!

**Comment by inverse** [comment]
The 40 character limit could be over come by using their API since we have the track ID. But the other issues will cause a problem. Having this on the backlog is a good idea in the mean time though :)

**Comment by HumbleBeeBumbleBee** [comment]
https://meta.stackexchange.com/questions/257547/scrobble-support-in-soundcloud-official-app

**Comment by HumbleBeeBumbleBee** [comment]
@inverse Any idea how we could implement some code to grab detailed track information from the SoundCloud API?

**Comment by HumbleBeeBumbleBee** [comment]
>
Thanks for getting in touch! We've received a lot of requests for this feature and our developers are currently looking into the infrastructure to make this possible, but they haven't provided me with a definitive answer as to when this might become available. Keep up with our status blog (status.soundcloud.com) to see any new updates and releases.

>I've forwarded your request to our developers to review for future updates and to emphasize the growing desire for this feature.

>Hope this helps! Let me know if you have any other questions.

That's seems promising!

**Comment by bhb192** [comment]
It sure does! Also I forgot to mention that since I've been using the stock
Music app, I've been having this problem where if I've opened SoundCloud at
all since the last reboot, it will automatically launch and start playing
upon Bluetooth Audio connection.

Even if I have already force stopped it and cleared the cache. I can even
have the stock Music app open and playing, and SoundCloud will open itself,
stop the stock app, and start playing I guess the last song I was listening
to.

Could this be part of the problem that you're having with the receiver for
SoundCloud also interfering with other apps?

**Comment by HumbleBeeBumbleBee** [comment]
@bhb192 not sure what could be causing that. Maybe it's SoundCloud. I have the same problem with Chrome and Google Play opening sometimes.

**Comment by HumbleBeeBumbleBee** [comment]
SLS accepts broadcasts for SoundCloud now.

**Comment by fishstik** [comment]
Why is this closed? I don't see SC on the apps list. I noticed the official Last.fm scrobbles SC now...though it looks like they're also restricted by the char limit. I've switched over to the official app just for this reason.

If we could add SC scrobbling support with some string parsing of the title (a lot of SC titles are 'Artist - Title'), I'd switch back, because SLS seems more stable overall.
